File: Data Structures & Algorithms/products-of-array-discluding-self/submission-0.js
class Solution {
    /**
     * @param {number[]} nums
     * @return {number[]}
     */
    productExceptSelf(nums) {
        let countZero = 0;
        const product = nums.reduce((prev, curr) => {
            if(curr == 0) {
                countZero += 1;
                return countZero > 1 ? 0 : prev;
            }
            return prev*curr;;
        }, 1);
        if (countZero > 1) {
            return new Array(nums.length).fill(0);
        }
        if (countZero === 1) {
            let zeroIndex = nums.indexOf(0);
            let output = new Array(nums.length).fill(0);
            output[zeroIndex] = product;
            return output;
        }
        return nums.map(num => product / num);
    }
}

File: Data Structures & Algorithms/subsets/submission-0.py
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        output = []
        def generate_subset(i, curr):
            if i == len(nums):
                output.append(curr.copy())
                return
            curr.append(nums[i])
            generate_subset(i + 1, curr)
            curr.pop()
            generate_subset(i + 1, curr)
        generate_subset(0, [])
        return output
